theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -319,19 +398,67 @@ def c_code(self, node, name, inputs, outputs, sub):
                                      PyArray_NDIM(%(name)s_tmp),
                                      (size_t *)PyArray_DIMS(%(name)s_tmp),
                                      (ssize_t *)PyArray_STRIDES(%(name)s_tmp),
-                                     pygpu_default_context(),
+                                     %(ctx)s,
                                      Py_None);
-
         Py_DECREF(%(name)s_tmp);
-        if (%(out)s == NULL)
-          %(fail)s
-        """ % {'name': name, 'inp': inputs[0],
+        if (%(out)s == NULL) {
+            %(fail)s
+        }
+        """ % {'name': name, 'inp': inputs[0], 'ctx': sub['context'],
                'out': outputs[0], 'fail': sub['fail']}
 
     def c_code_cache_version(self):
-        return (5,)
+        return (7,)
+
 
-gpu_from_host = GpuFromHost()
+class GpuToGpu(Op):
+    __props__ = ('context_name',)
+    _f16_ok = True
+    context_type = gpu_context_type
+
+    def __init__(self, context_name):
+        self.context_name = context_name
+
+    def __str__(self):
+        return 'GpuToGpu<%s>' % (self.context_name,)
+
+    def make_node(self, x):
+        if not isinstance(x.type, GpuArrayType):
+            raise TypeError(x)
+        return Apply(self, [x], [GpuArrayType(broadcastable=x.broadcastable,
+                                              context_name=self.context_name,
+                                              dtype=x.dtype)()])
+
+    def get_context(self, node):
+        return get_context(self.context_name)
+
+    def perform(self, node, inp, out, ctx):
+        x, = inp
+        z, = out
+        z[0] = x.transfer(ctx)
+
+    def grad(self, inputs, grads):
+        gz, = grads
+        return [GpuToGpu(inputs[0].type.context_name)(gz)]
+
+    def R_op(self, inputs, eval_points):
+        return self(eval_points[0])
+
+    def infer_shape(self, node, xshp):
+        return xshp
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        return """
+        Py_XDECREF(%(out)s);
+        %(out)s = pygpu_transfer(%(inp)s, %(ctx)s, 0);
```

Comment:
It would be useful to make pygpu_transfer take an output array for the destination. This would make allow_gc=False Speed up things here.